### PR TITLE
Migrate to the new OpenAssetIO unified entity data model

### PR DESCRIPTION
**Pull request by foundrytom**
Draft till https://github.com/TheFoundryVisionmongers/OpenAssetIO-MediaCreation/pull/2, https://github.com/TheFoundryVisionmongers/OpenAssetIO-MediaCreation/pull/3 and https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/363 are merged.

Once in, this switches to using the new trait-centric model for resolution.

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -39,25 +38,55 @@ jobs:
...
-          repository: TheFoundryVisionmongers/OpenAssetIO
+          repository: foundrytom/OpenAssetIO
+          ref: work/347-pythonVenvBuild
           path: dependencies/OpenAssetIO
 
+      - name: Cache conan packages
+        id: cache-conan
+        uses: actions/cache@v2
+        with:
+          path: ~/conan
```
Comment (feltech) [inline .github/workflows/ci.yaml]: [fix] It's a bit confusing, but `conan install --install-folder ~/conan ...` only tells conan where to put generated config files (e.g. CMake scripts).  By default the actual packages are installed in `~/.conan` (note the `.`).  

In the Vagrant bootstrap script(s) in OpenAssetIO there is a line `export CONAN_USER_HOME="$HOME/conan"` - that's the bit that tells conan where to install packages to.  In OpenAssetIO the packages and generated scripts are put in the same location, simply for convenience.

So we have a few options. The simplest is to just change this line to 
```suggestion
          path: ~/.conan
```
It means the generated files are not cached, but then they're probably regenerated every time anyway.

Or we could set the `CONAN_USER_HOME` environment variable, like in OpenAssetIO, to be explicit.

File: README.md [hunk 2]
```diff
@@ -61,14 +61,19 @@ python -m venv .venv
...
+At present, this requires installation from source of the
+`feature/181-coreCppBuild` branch.
 
 ```shell
 mkdir dependencies
-git clone git@github.com:TheFoundryVisionmongers/OpenAssetIO.git dependencies/OpenAssetIO
-pip install dependencies/OpenAssetIO
+git clone -b feature/181-coreCppBuild git@github.com:TheFoundryVisionmongers/OpenAssetIO.git dependencies/OpenAssetIO
+cmake -S dependencies/OpenAssetIO -B build
+cmake --build build --target openassetio-python-venv
```
Comment (feltech) [inline README.md]: [fix] This branch needs updating
```suggestion
cmake --build build --target openassetio-python-py-install
```

File: README.md [hunk 3]
```diff
@@ -61,14 +61,19 @@ python -m venv .venv
 ### Dependencies
 
 To use the linker in OpenTimelineIO `openassetio` needs to be installed.
-At present, this requires installation from source:
+At present, this requires installation from source of the
+`feature/181-coreCppBuild` branch.
```
Comment (feltech) [inline README.md:67]: [fix] OpenAssetIO-MediaCreation is also a dependency now

File: otio_openassetio/operations/openassetio_media_linker.py [hunk 4]
```diff
@@ -36,19 +40,33 @@ def link_media_reference(in_clip, media_linker_argument_map):
...
+    # This is perhaps more illustrative than useful at this point.
     context = session_state.context
-    context.locale.clipName = in_clip.name
-
-    mr.target_url = extract_one(
-        manager.resolveEntityReference([mr.target_url], context)
+    ClipTrait(context.locale).setName(in_clip.name)
+
+    # In this simple implementation, we only need the URL to the media,
+    # so we use the LocatableContentTrait directly to avoid the manager
```
Comment (feltech) [inline otio_openassetio/operations/openassetio_media_linker.py]: [consider] Might be worth adding a note "as opposed to all the traits of a particular specification"?

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -39,25 +38,55 @@ jobs:
...
         run: |
+          . dist/bin/activate
           flake8 --show-source --statistics otio_openassetio tests
 
       - name: Test with pytest
         run: |
+          . dist/bin/activate
           pytest tests
         env:
           OPENASSETIO_PLUGIN_PATH: dependencies/OpenAssetIO/resources/examples/manager/BasicAssetLibrary/plugin
```
Comment (feltech) [inline .github/workflows/ci.yaml]: [question] Not part of this PR, but the `bal_plugin_env` pytest fixture overwrites the env var, so is this here for a particular reason (similarly in the README instructions)?

**Review by feltech — CHANGES_REQUESTED** [review]
Took me a while to wrangle things to get tests running, but looks good and works well.  Just a couple of CI/docs fixes.

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (foundrytom) [reply .github/workflows/ci.yaml]: Ahhhhhh, awesome thanks, I did wonder if that var was a DRY device, or actually also doing Conan things. I like your suggestion, given hopefully all this will go soon.

File: README.md (hunk 2 above)
Comment (foundrytom) [reply README.md]: 🦅 👁️

File: README.md (hunk 3 above)
Comment (foundrytom) [reply README.md:67]: Added [here](https://github.com/TheFoundryVisionmongers/otio-openassetio/pull/5/commits/e3e92937a87978e2b8e3a52db8401d846f17ab6c).

File: otio_openassetio/operations/openassetio_media_linker.py (hunk 4 above)
Comment (foundrytom) [reply otio_openassetio/operations/openassetio_media_linker.py]: Extended [here](https://github.com/TheFoundryVisionmongers/otio-openassetio/pull/5/commits/ae293a72c290d407199b44744ae4c4f7bf5fae5a), and further corrected [here](https://github.com/TheFoundryVisionmongers/otio-openassetio/pull/5/commits/400f641a4103b8357f49f6ac5ff4a0e64f23d3fb).

File: .github/workflows/ci.yaml (hunk 5 above)
Comment (foundrytom) [reply .github/workflows/ci.yaml]: You are 100% right there good sir. Addressed [here](https://github.com/TheFoundryVisionmongers/otio-openassetio/pull/5/commits/96c5e22038d08d5c8ad123294f31dc109954d025).

File: README.md [hunk 6]
```diff
@@ -60,15 +60,29 @@ python -m venv .venv
...
 mkdir dependencies
-git clone git@github.com:TheFoundryVisionmongers/OpenAssetIO.git dependencies/OpenAssetIO
-pip install dependencies/OpenAssetIO
+git clone -b feature/181-coreCppBuild git@github.com:TheFoundryVisionmongers/OpenAssetIO.git dependencies/OpenAssetIO
+cmake -S dependencies/OpenAssetIO -B build
+cmake --build build --target openassetio-python-py-install
+. dist/bin activate
 ```
 
+The Media Creation library is more straight forward.
```
Comment (feltech) [inline README.md]: [consider]
```suggestion
The Media Creation library is more straightforward.
```

**Review by feltech — APPROVED** [review]
Just a minor typo, feel free to ignore/fix then squash and merge.
